File: graphs/kahns.js
function topologicalSort(graph) {
    // Count the incoming edges for each vertex
    const inDegree = new Map();
    for (const vertex in graph) {
      inDegree.set(vertex, 0);
    }
    for (const vertices of Object.values(graph)) {
      for (const vertex of vertices) {
        inDegree.set(vertex, inDegree.get(vertex) + 1);
      }
    }
  
    // Initialize a queue with vertices having no incoming edges
    const queue = [];
    for (const [vertex, degree] of inDegree.entries()) {
      if (degree === 0) {
        queue.push(vertex);
      }
    }
  
    // Perform topological sorting
    const result = [];
    while (queue.length > 0) {
      const current = queue.shift();
      result.push(current);
      for (const neighbor of graph[current]) {
        inDegree.set(neighbor, inDegree.get(neighbor) - 1);
        if (inDegree.get(neighbor) === 0) {
          queue.push(neighbor);
        }
      }
    }
  
    // Check for cycles
    if (result.length !== Object.keys(graph).length) {
      throw new Error("The graph contains a cycle!");
    }
  
    return result;
  }
  
  // Example usage:
  const graph = {
    "A": ["B", "C"],
    "B": ["D"],
    "C": ["D"],
    "D": []
  };
  
  console.log(topologicalSort(graph)); // Output: [ 'A', 'C', 'B', 'D' ]
  

File: graphs/kahns.py
from collections import deque


def topological_sort(vertices, edges):
    """
    Perform topological sorting on a directed acyclic graph (DAG).

    Args:
        vertices (list): A list of vertices in the graph.
        edges (list): A list of edges in the graph represented as tuples (u, v),
                      where u and v are vertices and u points to v.

    Returns:
        list: A list representing the topological order of the vertices.
              If the graph has a cycle, an empty list is returned.
    """
    in_degree = {v: 0 for v in vertices}
    adjacency_list = {v: [] for v in vertices}

    for u, v in edges:
        adjacency_list[u].append(v)
        in_degree[v] += 1

    queue = deque([v for v in vertices if in_degree[v] == 0])
    topological_order = []

    while queue:
        node = queue.popleft()
        topological_order.append(node)

        for neighbor in adjacency_list[node]:
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)

    if len(topological_order) == len(vertices):
        return topological_order
    else:
        return []  # Graph has a cycle

# Example usage:
vertices = [0, 1, 2, 3, 4, 5]
edges = [(5, 2), (5, 0), (4, 0), (4, 1), (2, 3), (3, 1)]
print(topological_sort(vertices, edges))  # Output: [4, 5, 2, 0, 3, 1]
